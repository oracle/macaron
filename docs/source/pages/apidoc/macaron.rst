.. Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
.. Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

macaron package
===============

.. automodule:: macaron
   :members:
   :undoc-members:
   :show-inheritance:

Subpackages
-----------

.. toctree::
   :maxdepth: 1

   macaron.code_analyzer
   macaron.config
   macaron.database
   macaron.dependency_analyzer
   macaron.output_reporter
   macaron.parsers
   macaron.policy_engine
   macaron.slsa_analyzer

Submodules
----------

macaron.util module
-------------------

.. automodule:: macaron.util
   :members:
   :undoc-members:
   :show-inheritance:
