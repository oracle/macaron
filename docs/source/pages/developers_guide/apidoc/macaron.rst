macaron package
===============

.. automodule:: macaron
   :members:
   :undoc-members:
   :show-inheritance:

Subpackages
-----------

.. toctree::
   :maxdepth: 1

   macaron.code_analyzer
   macaron.config
   macaron.database
   macaron.dependency_analyzer
   macaron.output_reporter
   macaron.parsers
   macaron.policy_engine
   macaron.slsa_analyzer

Submodules
----------

macaron.errors module
---------------------

.. automodule:: macaron.errors
   :members:
   :undoc-members:
   :show-inheritance:

macaron.util module
-------------------

.. automodule:: macaron.util
   :members:
   :undoc-members:
   :show-inheritance:
