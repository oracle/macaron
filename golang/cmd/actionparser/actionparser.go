/* Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved. */
/* Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/. */

package main

import (
	"flag"
	"fmt"
	"os"

	"github.com/oracle-samples/macaron/golang/internal/actionparser"
	"github.com/oracle-samples/macaron/golang/internal/filewriter"
)

// Parse the content of a GitHub Action yaml file and
// print the JSON format for it to stdout. It can also store the JSON content to a file.
// Params:
//
//	-file <FILE_PATH>: the path to the GitHub Actions file
//	-output <OUTPUT_FILE>: the output file path to store the JSON content.
//
// Return code:
//
//	0 - Parse successfully, return the JSON as string to stdout. If -output is set, store the json content to the file.
//		If there is any error storing to file, the result is still put to stdout, but the errors are put to stderr instead.
//	1 - Error: Missing workflow file path.
//	2 - Error: Could not parse the GitHub Action workflow files. Parse errors will be printed to stderr.
func main() {
	file_path := flag.String("file", "", "The path of the GitHub Actions workflow file.")
	out_path := flag.String("output", "", "The output file path to store the JSON content.")
	flag.Parse()

	if len(*file_path) <= 0 {
		fmt.Fprintln(os.Stderr, "Missing workflow file path.")
		flag.PrintDefaults()
		os.Exit(1)
	}

	json_content, err := actionparser.ParseActionFromFile(*file_path)
	if err == nil {
		fmt.Fprintln(os.Stdout, json_content)
		if len(*out_path) > 0 {
			err := filewriter.StoreBytesToFile([]byte(json_content), *out_path)
			if err != nil {
				fmt.Fprintln(os.Stderr, err.Error())
			}
		}
		os.Exit(0)
	} else {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(2)
	}
}
