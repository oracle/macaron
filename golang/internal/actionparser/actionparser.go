/* Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved. */
/* Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/. */

// Package actionparser parses GitHub Actions and provides the parsed object in JSON.
package actionparser

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"

	"github.com/rhysd/actionlint"
)

// printParseErrs prints the parse errors from actionlint to w.
func printParseErrs(w io.Writer, file_path string, errs []*actionlint.Error) {
	for _, err := range errs {
		err.Filepath = file_path
		fmt.Fprintln(w, err.Error())
	}
}

// ParseActionFromFile parses the content of the given yaml file.
// It returns the parsed object in JSON format.
func ParseActionFromFile(file string) (string, error) {
	data, read_err := os.ReadFile(file)
	if read_err != nil {
		return "", read_err
	}

	workflow, parse_err := actionlint.Parse(data)
	if parse_err != nil {
		printParseErrs(os.Stderr, file, parse_err)
		return "", errors.New("parse: errors while parsing the gitHub actions yaml file")
	}

	result, json_err := json.MarshalIndent(workflow, "", "  ")
	if json_err != nil {
		return "", json_err
	}

	return string(result), nil
}
