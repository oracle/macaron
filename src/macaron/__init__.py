# Copyright (c) 2022 - 2023, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module initializes the necessary components for the macaron package."""

import os

# The version of this package. There's no comprehensive, official list of other
# magic constants, so we stick with this one only for now. See also this conversation:
# https://stackoverflow.com/questions/38344848/is-there-a-comprehensive-table-of-pythons-magic-constants
__version__ = "0.0.0"

# The path to the Macaron package.
MACARON_PATH = os.path.dirname(os.path.abspath(__file__))
