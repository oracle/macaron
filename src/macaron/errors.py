# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains error classes for Macaron."""


class MacaronError(Exception):
    """The base class for Macaron errors."""


class InvalidExpectationError(MacaronError):
    """Happens when the provenance expectation is invalid."""


class ExpectationRuntimeError(MacaronError):
    """Happens if there are errors while validating the expectation against a target."""


class CUEExpectationError(MacaronError):
    """Happens when the CUE expectation is invalid."""


class CUERuntimeError(MacaronError):
    """Happens when there are errors in CUE expectation validation."""


class ConfigurationError(MacaronError):
    """Happens when there is an error in the configuration (.ini) file."""


class CloneError(MacaronError):
    """Happens when cannot clone a git repository."""


class RepoCheckOutError(MacaronError):
    """Happens when there is an error when checking out the correct revision of a git repository."""


class RepoNotFoundError(MacaronError):
    """Happens if a repository is not found."""


class PURLNotFoundError(MacaronError):
    """Happens if the PURL identifier for a software component is not found."""


class DuplicateError(MacaronError):
    """The class for errors for duplicated data."""
