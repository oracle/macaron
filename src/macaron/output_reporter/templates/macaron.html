{#
    Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
    Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
#}

{% extends "base_template.html"%}
{# This section contains the reuseable macros. #}

{#
    This macro renders a hyperlink.
#}
{%- macro render_hyperlink(href, display_text= "", target="_blank", class="", id="") -%}
<a href="{{ href }}" target="{{ target }}" class="{{ class }}" id = "{{ id }}">{{ display_text }}</a>
{%- endmacro %}

{#
    This macro creates the <thead> element for a table from a list of header names.
#}
{%- macro render_headers(headers) -%}
<thead>
    <tr>
        {% for header in headers %}
        <th>{{ header | replace("_", " ") | capitalize }}</th>
        {% endfor %}
    </tr>
</thead>
{%- endmacro %}

{#
    This macro renders the structure for a html table which will contain <thead> (optional) and <tbody> elements.
    The parameter header_names is a list of header names for the table (default to an empty list).
    If has_header == true, the macro will also render the from header_names (default to true).
    The rendering for each data row in the table must be handled by the caller.
#}
{%- macro render_table(header_names=[], has_header=true, class="", id="") -%}
<table class="{{ class }}" id="{{ id }}">
    {% if has_header == true %}
    {%- set headers = header_names -%}
    {{ render_headers(headers) | indent(4, first=true) }}
    {% endif %}
    <tbody>
    {# We use the caller rendering here. #}
    {{ caller() | indent(4) }}
    </tbody>
</table>
{%- endmacro %}

{#
    This macro renders a hierarchy list to display a dictionary-structured data.
#}
{% macro render_tree_view_nested_list(data, class="", id="") %}
{% set data = data | get_flatten_dict %}
<ul class="tree-view-nested-list {{ class }}" id="{{ id }}">
    {% for key, value in data.items() recursive %}
    <li>
        {% if value is mapping %}
        <div class="caret">{{ key }}:</div>
        <ul class="nested-ul">
        {{ loop(value.items()) | indent(2 * 4) }}
        </ul>
        {% else %}
        {{ key }}: {{ value }}
        {% endif %}
    </li>
    {% endfor %}
</ul>
{% endmacro %}

{# -------------------------------------------- #}

{#
    This section contains macros that are use to render each element in this template.
    We define these macros separately so that we could use correct
    indentation for the whole block.
#}

{% macro render_target_info_table(target_info_data) %}
{% call render_table(class="target_info", has_header=false) %}
    {% for key, value in target_info_data | items %}
    <tr>
        <th>
            {{ key | replace("_", " ") | capitalize }}
        </th>
        <td>
            {{ value }}
        </td>
    </tr>
    {% endfor %}
{% endcall %}
{% endmacro %}

{% macro render_checks_report(checks_report_data) %}
{%- set headers = checks_report_data | get_headers -%}
{% call render_table(header_names=headers, has_header=true, class="checks_report") %}
    {% for item in checks_report_data %}
    <tr>
    {% for header in headers %}
        {% if header == "slsa_requirements" %}
        <td>
            <ul>
            {% for ele in item[header] %}
                <li>{{ ele }}</li>
            {% endfor %}
            </ul>
        </td>
        {% elif header == "result_type" %}
        <td class={{ item[header] | get_check_result_color }}>{{ item[header].value }}</td>
        {% elif header == "justification" %}
        <td>
            <ul>
            {% for ele in item[header] %}
                {% if ele is mapping %}
                    {% for key, value in ele | items %}
                <li>{{ key }}: {{ render_hyperlink(href=value, display_text=value) }}</li>
                    {% endfor %}
                {% else %}
                <li>{{ ele }}</li>
                {% endif %}
            {% endfor %}
            </ul>
        </td>
        {% else %}
        <td>{{ item[header] }}</td>
        {% endif %}
    {% endfor %}
    </tr>
    {% endfor %}
{% endcall %}
{% endmacro %}

{% macro render_dep_summary(dep_summary_data) %}
{%- set headers = dep_summary_data | get_headers -%}
{% call render_table(header_names=headers, has_header=true, class="dep_summary") %}
    {% for item in dep_summary_data %}
    <tr>
        {% for header in headers %}
        <td>{{ item[header] }}</td>
        {% endfor %}
    </tr>
    {% endfor %}
{% endcall %}
{% endmacro %}

{% macro render_dep_status(dep_status_data) %}
{%- set headers = dep_status_data | get_headers -%}
{% call render_table(header_names=headers, has_header=true, class="dep_status") %}
    {% for item in dep_status_data %}
    <tr>
    {% for header in headers %}
        {% if header == "report" %}
        {% if item[header] %}
        <td>{{ render_hyperlink(item[header], display_text="Open report.") }}</td>
        {% else %}
        <td>Not available.</td>
        {% endif %}
        {% elif header == "status" %}
        <td class={{ item[header] | get_dep_status_color }}>{{ item[header].value }}</td>
        {% else %}
        <td>{{ item[header] }}</td>
        {% endif %}
    {% endfor %}
    </tr>
    {% endfor %}
{% endcall %}
{% endmacro %}

{# -------------------------------------------- #}

{#
    This section is where we fill in the blocks defined in base_template.html.
    Note that the input context dictionary will have the format:
        {data = <data_from_html_reporter}
#}

{% block timestamps %}
{{ metadata.timestamps | indent(12, first=true) }}
{% endblock%}


{% block target_info %}
{{ render_target_info_table(target.info) | indent(4, first=true) }}
{% endblock %}


{% block prov_justification %}
    <div id= "prov_justification">
        {% if target.provenances.is_inferred == true %}
        Could not find a provenance for this repository. Below is what Macaron has inferred.
        {% else %}
        This is the provenance found for this repository.
        {% endif %}
    </div>
{% endblock %}


{% block prov_summary %}
{{ render_tree_view_nested_list(target.provenances.content) | indent(4, first=true) }}
{% endblock %}


{% block checks_report %}
{{ render_checks_report(target.checks.results) | indent(4, first=true) }}
{% endblock %}


{% block deps_result_section %}
{% if dependencies.analyzed_deps != 0 %}
    <div class="table_caption" id="deps_result_title">Dependency results</div>
    <div class="table_sub_caption">
    {{ dependencies.analyzed_deps }} dependencies that map to {{ dependencies.unique_dep_repos }} unique repositories have been successfully analyzed.
    </div>
    {{ render_dep_summary(dependencies.checks_summary) | indent(4) }}
    <div class="space_divider"></div>
    {{ render_dep_status(dependencies.dep_status) | indent(4) }}
{% endif %}
{% endblock %}
