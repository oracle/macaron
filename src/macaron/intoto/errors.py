# Copyright (c) 2023 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""Error types related to in-toto attestations."""

from macaron.errors import MacaronError


class InTotoAttestationError(MacaronError):
    """The base error type for all in-toto related errors."""


class ValidateInTotoPayloadError(InTotoAttestationError):
    """Happens when there is an issue validating an in-toto payload, usually against a schema."""


class UnsupportedInTotoVersionError(InTotoAttestationError):
    """Happens when encountering an attestation under an unsupported in-toto version."""


class DecodeIntotoAttestationError(InTotoAttestationError):
    """Happens when there is an issue decoding the payload of an in-toto attestation."""


class LoadIntotoAttestationError(InTotoAttestationError):
    """Happens when there is an issue loading the payload of an in-toto attestation."""
