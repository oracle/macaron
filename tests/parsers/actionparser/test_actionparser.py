# Copyright (c) 2022 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""
This module tests the GitHub Actions parser.
"""

import os
from pathlib import Path

import pytest

from macaron import MACARON_PATH
from macaron.errors import ParseError
from macaron.parsers.actionparser import parse

RESOURCES_DIR = Path(__file__).parent.joinpath("resources")


@pytest.mark.parametrize(
    "workflow_path",
    [
        "codeql-analysis.yml",
        "maven.yml",
    ],
)
def test_actionparser_parse(snapshot: dict, workflow_path: str) -> None:
    """Test parsing GH Actions workflows."""
    assert parse(os.path.join(RESOURCES_DIR, "workflow_files", workflow_path), MACARON_PATH) == snapshot


@pytest.mark.parametrize(
    "workflow_path",
    [
        "release.yml",
        "invalid.yml",
        "file_does_not_exist.yml",
    ],
)
def test_actionparser_parse_invalid(workflow_path: str) -> None:
    """Test parsing GH Actions workflows."""
    with pytest.raises(ParseError):
        parse(os.path.join(RESOURCES_DIR, "workflow_files", workflow_path), MACARON_PATH)
