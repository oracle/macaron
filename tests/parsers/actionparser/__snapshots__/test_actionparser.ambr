# serializer version: 1
# name: test_actionparser_parse[codeql-analysis.yml]
  dict({
    'jobs': dict({
      'analyze': dict({
        'name': 'Analyze',
        'permissions': dict({
          'actions': 'read',
          'contents': 'read',
          'security-events': 'write',
        }),
        'runs-on': 'ubuntu-latest',
        'steps': list([
          dict({
            'name': 'Checkout repository',
            'uses': 'actions/checkout@v2',
          }),
          dict({
            'name': 'Initialize CodeQL',
            'uses': 'github/codeql-action/init@v2',
            'with': dict({
              'languages': '${{ matrix.language }}',
            }),
          }),
          dict({
            'name': 'Autobuild',
            'uses': 'github/codeql-action/autobuild@v2',
          }),
          dict({
            'name': 'Perform CodeQL Analysis',
            'uses': 'github/codeql-action/analyze@v2',
          }),
        ]),
        'strategy': dict({
          'fail-fast': False,
          'matrix': dict({
            'language': list([
              'java',
            ]),
          }),
        }),
      }),
    }),
    'name': 'CodeQL',
    'on': dict({
      'pull_request': dict({
        'branches': list([
          'master',
          '3.0',
          '2.14',
          '2.13',
        ]),
        'paths-ignore': list([
          'README.md',
          'release-notes/*',
        ]),
      }),
      'push': dict({
        'branches': list([
          'master',
          '3.0',
          '2.14',
          '2.13',
        ]),
        'paths-ignore': list([
          'README.md',
          'release-notes/*',
        ]),
      }),
      'schedule': list([
        dict({
          'cron': '40 21 * * 2',
        }),
      ]),
    }),
  })
# ---
# name: test_actionparser_parse[maven.yml]
  dict({
    'jobs': dict({
      'build': dict({
        'runs-on': '${{ matrix.os }}',
        'steps': list([
          dict({
            'uses': 'actions/checkout@v2',
          }),
          dict({
            'uses': 'actions/setup-java@v2',
            'with': dict({
              'cache': 'maven',
              'distribution': 'temurin',
              'java-version': 8,
            }),
          }),
          dict({
            'name': 'Build with Maven',
            'run': 'mvn verify -e -B -V -DdistributionFileName=apache-maven',
          }),
          dict({
            'if': "${{ matrix.os == 'ubuntu-latest' }}",
            'name': 'Upload built Maven',
            'uses': 'actions/upload-artifact@v2',
            'with': dict({
              'name': 'built-maven',
              'path': 'apache-maven/target/',
            }),
          }),
        ]),
        'strategy': dict({
          'fail-fast': False,
          'matrix': dict({
            'os': list([
              'ubuntu-latest',
              'windows-latest',
              'macOS-latest',
            ]),
          }),
        }),
      }),
      'integration-test': dict({
        'needs': 'build',
        'runs-on': '${{ matrix.os }}',
        'steps': list([
          dict({
            'env': dict({
              'PR_HEAD_LABEL': '${{ github.event.pull_request.head.label }}',
            }),
            'name': 'Collect environment context variables',
            'run': '''
              set +e
              repo=maven-integration-testing
              target_branch=master
              target_user=apache
              if [ "$GITHUB_EVENT_NAME" == "pull_request" ]; then
                user=${PR_HEAD_LABEL%:*}
                branch=${PR_HEAD_LABEL#*:}
              else
                user=${GITHUB_REPOSITORY%/*}
                branch=${GITHUB_REF#refs/heads/}
              fi
              if [ $branch != "master" ]; then
                git ls-remote https://github.com/$user/$repo.git | grep "refs/heads/${branch}$" > /dev/null
                if [ $? -eq 0 ]; then
                  echo "Found a branch \"$branch\" in fork \"$user/$repo\", configuring this for the integration tests to be run against."
                  target_branch=$branch
                  target_user=$user
                else
                  echo "Could not find fork \"$user/$repo\" or a branch \"$branch\" in this fork. Falling back to \"$target_branch\" in \"$target_user/$repo\"."
                fi
              else
                echo "Integration tests will run against $target_user/$repo for master builds."
              fi
              echo "REPO_BRANCH=$target_branch" >> $GITHUB_ENV
              echo "REPO_USER=$target_user" >> $GITHUB_ENV
  
            ''',
            'shell': 'bash',
          }),
          dict({
            'name': 'Checkout maven-integration-testing',
            'uses': 'actions/checkout@v2',
            'with': dict({
              'path': 'maven-integration-testing/',
              'ref': '${{ env.REPO_BRANCH }}',
              'repository': '${{ env.REPO_USER }}/maven-integration-testing',
            }),
          }),
          dict({
            'name': 'Set up cache for ~/.m2/repository',
            'uses': 'actions/cache@v2',
            'with': dict({
              'key': "it-m2-repo-${{ matrix.os }}-${{ hashFiles('maven-integration-testing/**/pom.xml') }}",
              'path': '~/.m2/repository',
              'restore-keys': '''
                it-m2-repo-${{ matrix.os }}-
  
              ''',
            }),
          }),
          dict({
            'name': 'Download built Maven',
            'uses': 'actions/download-artifact@v2',
            'with': dict({
              'name': 'built-maven',
              'path': 'built-maven/',
            }),
          }),
          dict({
            'name': 'Set up JDK',
            'uses': 'actions/setup-java@v2',
            'with': dict({
              'cache': 'maven',
              'distribution': 'temurin',
              'java-version': '${{ matrix.java }}',
            }),
          }),
          dict({
            'name': 'Running integration tests',
            'run': 'mvn install -e -B -V -Prun-its,embedded -Dmaven.repo.local="$HOME/.m2/repository" -DmavenDistro="$GITHUB_WORKSPACE/built-maven/apache-maven-bin.zip" -f maven-integration-testing/pom.xml',
            'shell': 'bash',
          }),
        ]),
        'strategy': dict({
          'fail-fast': False,
          'matrix': dict({
            'java': list([
              8,
              11,
              17,
            ]),
            'os': list([
              'ubuntu-latest',
              'windows-latest',
              'macOS-latest',
            ]),
          }),
        }),
      }),
    }),
    'name': 'Java CI',
    'on': list([
      'push',
      'pull_request',
    ]),
  })
# ---
